### PR TITLE
Fix Juno R2 energy model

### PR DESCRIPTION
**Pull request by douglas-raillard-arm**
1) Fix Juno R2 energy model.
2) Change the optimal task placement algorithm used by LISA to better match EAS behavior.
EDIT: removed 2)

### CODE REVIEW

**Comment by qperret** [comment]
Since we're doing this it might be worth adding the juno r2 model to `libs/utils/platforms/juno_energy.py`.

File: src/shell/lisa_shell [hunk 1]
```diff
@@ -318,14 +321,22 @@ EOF
...
-nosetests -v --nocapture --nologcapture \
-          --logging-config=logging.conf \
-          $*
+if [ -n "$LISA_RESULTS_DIR" ]; then
+  mkdir -p "$LISA_RESULTS_DIR"
+  # There is no sensible place to store results.xml when the directory is not
+  # set explicitely, as the Python scripts decide where to put the results..
+  # Using results_latest symlink is not an option as it would race if using the
+  # same tree to run multiple tests on different boards at the same time.
+  local xunit_options=(--with-xunit --xunit-file="$LISA_RESULTS_DIR/results.xml")
```
Comment (valschneider) [inline src/shell/lisa_shell]: Ah, that explains the environment variable. We could read the `results_dir` option from the file conf but that means reading JSON in bash...

File: src/shell/lisa_shell [hunk 2]
```diff
@@ -335,24 +346,37 @@ while [[ $# -gt 0 ]]; do
...
+                        shift
+                        ;;
+                -o|--results-dir)
+                        LISA_RESULTS_DIR="$2"
                         shift
                         ;;
                 *)
                         # Unrecognised args are passed through to nosetests
-                        extra_args="$extra_args $1"
+						extra_args+=("$1")
```
Comment (valschneider) [inline src/shell/lisa_shell]: That's more than one whitespace too far :p

File: libs/utils/env.py [hunk 3]
```diff
@@ -239,17 +248,27 @@ def __init__(self, target_conf=None, test_conf=None, wipe=True,
...
-                        self.conf.get('results_dir'))
-
-        if self.res_dir and not os.path.isabs(self.res_dir):
-            self.res_dir = os.path.join(basepath, 'results', self.res_dir)
-        else:
+        # Initialize local results folder.
+        # The test configuration overrides the target's one and the environment
+        # variable overrides everything else.
+        self.res_dir = (
+            os.getenv('LISA_RESULTS_DIR') or
```
[the remark this answers is not in the document]
Comment (douglas-raillard-arm) [reply libs/utils/env.py]: Setting this using the JSON file has 2 issues:
1) target.config being about the target, the results dir could very well be different when you test your kernel 1 and kernel 2
2) The test "engine" is split in 2 parts: TestEnv and plumbing classes like that, and nosetests itself. The only way to pass information to both (--xunit-file=<path>) is from the lisa_shell script. It is not straightforward to read target.json from lisa_shel without writing a python script to read the JSON and print the value back.

We are also already applying this pattern for --iteration option of lisa-test.

File: libs/utils/env.py [hunk 4]
```diff
@@ -239,17 +248,27 @@ def __init__(self, target_conf=None, test_conf=None, wipe=True,
...
+            self.conf.get('results_dir') or
+            # THIS MUST NOT BE USED. It is only available for a specific test,
+            # but must not be relied upon as the user is able to override that.
+            self.test_conf.get('results_dir')
```
[the remark this answers is not in the document]
Comment (douglas-raillard-arm) [reply libs/utils/env.py]: Actually, this is used in tests/eas/preliminary.py (TEST_CONF) , so more than one test.
It would be a good idea to update it, but it looks like it is not run on a regular basis (bring up test), so the pain is minimal here.

File: libs/utils/env.py [hunk 5]
```diff
@@ -239,17 +248,27 @@ def __init__(self, target_conf=None, test_conf=None, wipe=True,
...
+        )
+        if not self.res_dir:
+            # If the loop terminates without break, it means no res_dir has been
```
[the remark this answers is not in the document]
Comment (douglas-raillard-arm) [reply libs/utils/env.py]: Outdated comment, I rewrote this part indeed :)

File: libs/utils/test.py [hunk 6]
```diff
@@ -271,8 +271,10 @@ def experiment_test(wrapped_test, instance, args, kwargs):
 # runExperiments, so that if the value is invalid we print the error
 # immediately instead of going ahead with target setup etc.
 try:
-    ITERATIONS_FROM_CMDLINE = int(
-        os.getenv('LISA_TEST_ITERATIONS', '0'))
+    iterations = os.getenv('LISA_TEST_ITERATIONS')
```
[the remark this answers is not in the document]
Comment (douglas-raillard-arm) [reply libs/utils/test.py]: This was already there :) , the only change here is making the interface simpler (allowing empty variable)

File: src/shell/lisa_shell [hunk 7]
```diff
@@ -283,6 +284,8 @@ function _lisa-test-usage {
 cat <<EOF
 Usage: lisa-test [--iterations ITERATIONS] [args] FILE[:CLASS]
   Run automated tests. Tests can be found under the tests/ directory.
+  Only one python file can be specified at a time, otherwise the results might
```
[the remark this answers is not in the document]
Comment (douglas-raillard-arm) [reply src/shell/lisa_shell]: Yes, it is related to the wrong workload being executed for the 2nd file mentioned on the command line.

File: src/shell/lisa_shell (hunk 2 above)
Comment (douglas-raillard-arm) [reply src/shell/lisa_shell]: Ah, accidental conversion of spaces to tabs

**Comment by valschneider** [comment]
I'm ok with this split up as it is now. I would however like a more wordy comment for the 20% capacity thing - and ideally I'd like some other people's opinion on the matter (I have already expressed my thoughts). @derkling, @credp, @qperret  ?

**Comment by qperret** [comment]
Thinking about it I must say that I agree with @valschneider on this one. If the optimal placement is to use little CPUs at 99% of their capacity, then this is an issue in EAS and I'd like the tests to tell us that. If we have issues on Juno right now I vote for changing the size of the tasks in the test, and to remove the accounting of the 20% margin form the optimal task placement.

**Comment by douglas-raillard-arm** [comment]
@qperret : The choice we have to make is:
1) _guess_freqs computes the optimal task placement
2) _guess_freqs simulates the expected behavior of EAS

These are two different things, and the current code is a bit confusing since the functions are named as in 1) but are actually expected to behave like 2) by the rest of the code most of the time, since we compare the output of this function with what the actual EAS code did.

Maybe adding a parameter to these functions to know if we want the optimal placement or EAS expected behavior would allow tests to choose the behavior they need.

**Comment by qperret** [comment]
I vote for 1. The optimal placement should be just that: optimal. It should tell us the behaviour we should aim for, not mimic the behaviour we have. If we're too far off the optimal placement energy wise this is a good indication we have to fix something in EAS :)

**Comment by douglas-raillard-arm** [comment]
Ok, so do try and tweak the tests so that they never fall into the grey area were EAS and optimal placement disagree ? If we don't tweak, they will fail (but if we really want to check that EAS is strictly optimal, that is desired). If we do tweak, they will stay away from EAS limits, which means the test will not show small regressions.

If we mirror EAS behavior (via a parameter), we can check for actual regressions relative to the current EAS code, instead of an absolute "regression" compared to the optimal.

**Comment by douglas-raillard-arm** [comment]
Agreed on using the optimal placement and tweaking tests so they fall in the area where EAS is optimal, and testing the grey area behavior in different tests.

File: libs/utils/platforms/juno_energy.py [hunk 8]
```diff
@@ -90,7 +90,7 @@ def a57_cpu_node(cpu):
                            active_states=a57_cpu_active_states,
                            idle_states=a57_cpu_idle_states)
 
-juno_energy = EnergyModel(
+juno_r0_energy = EnergyModel(
```
Comment (derkling) [inline libs/utils/platforms/juno_energy.py]: Should we not rename the file too? That's for consistency as well as "to make it obvious" as you say in the commit message..

File: libs/utils/env.py [hunk 9]
```diff
@@ -410,50 +410,56 @@ def _init_target(self, force = False):
         if 'board' not in self.conf:
             self.conf['board'] = 'UNKNOWN'
 
+        board_name = self.conf['board'].upper()
```
Comment (derkling) [inline libs/utils/env.py]: I like this simplification... can we just move them into a dedicated separate patch?
Just to keep it more clean the history and better isolate the patch which change the JUNO energy model... you can keep the patch either before or after... whatever is more convenient.

File: libs/utils/env.py [hunk 10]
```diff
@@ -410,50 +410,56 @@ def _init_target(self, force = False):
...
+
         # Initialize TC2 board
-        if self.conf['board'].upper() == 'TC2':
+        if board_name == 'TC2':
             platform = devlib.platform.arm.TC2()
             self.__modules = ['bl', 'hwmon', 'cpufreq']
 
         # Initialize JUNO board
-        elif self.conf['board'].upper() in ('JUNO', 'JUNO2'):
+        elif board_name == 'JUNO':
```
Comment (derkling) [inline libs/utils/env.py]: Cannot we still try to keep one section per "board":
```python
 elif board_name in in ('JUNO', 'JUNO2'):
    # ...
    if board_name == JUNO:
        self.nrg_model = juno_r0_energy
    elif board_name == JUNO2:
        self.nrg_model = juno_energy
    # ...
```

File: libs/utils/env.py (hunk 9 above)
Comment (douglas-raillard-arm) [reply libs/utils/env.py]: Done.

File: libs/utils/env.py (hunk 10 above)
Comment (douglas-raillard-arm) [reply libs/utils/env.py]: Done.

File: libs/utils/platforms/juno_energy.py (hunk 8 above)
Comment (douglas-raillard-arm) [reply libs/utils/platforms/juno_energy.py]: Done.

File: libs/utils/env.py [hunk 11]
```diff
@@ -420,9 +420,11 @@ def _init_target(self, force = False):
         # Initialize JUNO board
         elif board_name in ('JUNO', 'JUNO2'):
             platform = devlib.platform.arm.Juno()
-            self.nrg_model = juno_energy
             self.__modules = ['bl', 'hwmon', 'cpufreq']
 
+            if board_name == 'JUNO':
+                self.nrg_model = juno_r0_energy
```
Comment (derkling) [inline libs/utils/env.py]: Aren't we now missing and nrg_model for `JUNO2`?
I.e. look at my previous proposal...

Comment (douglas-raillard-arm) [reply libs/utils/env.py]: It is not missing more than it used to be :-) Since I don't need it at the moment (I always extract it from the target), I'll leave that for another PR and for the person that actually needs the feature.

Comment (derkling) [reply libs/utils/env.py]: Ok, I see what you mean... and at the end that's exactly the same behaviour we had initially... actually, better, since before we was loading a wrong EM for JunoR2.

One last question, what happen on R0, do we still override the above hardcoded EM if we are able to read them from the target?

Comment (douglas-raillard-arm) [reply libs/utils/env.py]: That's right. Later in the same function, there is this bit of code, so the energy model is only read from the target if not already provided:
```python
        if not self.nrg_model:
            try:
                self._log.info('Attempting to read energy model from target')
                self.nrg_model = EnergyModel.from_target(self.target)
```
